src/borg/helpers/shellpattern.py

```diff
@@ -60,7 +61,86 @@ def translate(pat, match_end=r"\Z"):
                 elif stuff[0] == "^":
                     stuff = "\\" + stuff
                 res += "[%s]" % stuff
+        elif c == "(" or c == "|" or c == ")":
+            if i > 0 and pat[i - 1] != "\\":
+                res += c
         else:
             res += re.escape(c)
 
     return "(?ms)" + res + match_end
+
+
+def _parse_braces(pat):
+    """Returns the index values of paired braces in `pat` as <key, value> pairs, else None.
+
+    The dict's keys are the indexes corresponding to opening braces. Initially,
+    they are set to a value of -1. Once a corresponding closing brace is found,
+    the value is updated. All dict keys with a positive int value are valid pairs.
+
+    Cannot rely on re.match("[^\\(\\\\)*]?{.*[^\\(\\\\)*]}") because, while it
+    does handle unpaired braces and nested pairs of braces, it misses sequences
+    of paired braces. E.g.: "{foo,bar}{bar,baz}" would translate, incorrectly, to
+    "(foo|bar\\}\\{bar|baz)" instead of, correctly, to "(foo|bar)(bar|baz)"
+
+    So this function parses in a left-to-right fashion, tracking pairs with a LIFO
+    queue: pushing opening braces on and popping them off when finding a closing
+    brace.
+    """
+    curly_q = LifoQueue()
+    pairs: dict[int, int] = dict()
+
+    for idx, c in enumerate(pat):
+        if c == "{":
+            if idx == 0 or pat[idx - 1] != "\\":
+                # Opening brace is not escaped.
+                # Add to dict
+                pairs[idx] = -1
+                # Add to queue
+                curly_q.put(idx)
+        if c == "}" and curly_q.qsize():
+            # If queue is empty, then cannot close pair.
+            if idx > 0 and pat[idx - 1] != "\\":
+                # Closing brace is not escaped.
+                # Pop off the index of the corresponding opening brace, which
+                # provides the key in the dict of pairs, and set its value.
+                pairs[curly_q.get()] = idx
+    return pairs
```

Comment:
side note: guess having unmatched opening positions in the returned data isn't needed - it does not contain the position of unmatched closing positions either.

Reply:
I like this a lot. Initially I was thinking of returning a list of tuples. I'll just change the conditional to `if closing > 0`, since there's no way for a closing brace to be in the zero-index position. I will also change the -1 initiaul value for the closing index to 0 to reflect this.

Reply:
Well, I liked the -1 better as it is an "impossible index" and does not need any further thinking.

`None` might be an even better value (like "we do not know yet").